File: packages/folio-core/domain/issue.go
package domain

import "time"

type IssueKind string

const (
	IssueTicket IssueKind = "ticket"
	IssueTodo   IssueKind = "todo"
)

type Priority string

const (
	PriorityLow    Priority = "low"
	PriorityMedium Priority = "medium"
	PriorityHigh   Priority = "high"
)

type WayfinderType string

const (
	WayfinderMap       WayfinderType = "map"
	WayfinderResearch  WayfinderType = "research"
	WayfinderPrototype WayfinderType = "prototype"
	WayfinderGrilling  WayfinderType = "grilling"
	WayfinderTask      WayfinderType = "task"
)

type IssueStatus string

const (
	IssueOpen       IssueStatus = "open"
	IssueInProgress IssueStatus = "in_progress"
	IssueBlocked    IssueStatus = "blocked"
	IssueDone       IssueStatus = "done"
	IssueCancelled  IssueStatus = "cancelled"
)

func (s IssueStatus) IsTerminal() bool {
	return s == IssueDone || s == IssueCancelled
}

type Size int

const (
	SizeXS Size = 1
	SizeS  Size = 2
	SizeM  Size = 3
	SizeL  Size = 5
	SizeXL Size = 8
)

var sizes = map[Size]bool{SizeXS: true, SizeS: true, SizeM: true, SizeL: true, SizeXL: true}

func (s Size) Valid() bool { return sizes[s] }

type LinkKind string

const (
	LinkBlocks  LinkKind = "blocks"
	LinkRelates LinkKind = "relates"
	LinkParent  LinkKind = "parent"
)

// From is the subject: for LinkParent it is the child, for LinkBlocks the blocker.
type IssueLink struct {
	ID       string
	DomainID DomainID
	From     IssueID
	To       IssueID
	Kind     LinkKind
}

type Issue struct {
	ID          IssueID
	Kind        IssueKind
	ProjectID   ProjectID
	PlanID      PlanID
	Slug        string
	Title       string
	Body        string
	Status      IssueStatus
	Priority    Priority
	Size        Size
	Assignee    UserID
	Tags        []string
	Position    int
	DueDate     *time.Time
	Wayfinder   WayfinderType
	ExternalRef string
	CreatedBy   UserID
	CreatedAt   time.Time
	UpdatedAt   time.Time

	// Derived on read, never persisted.
	ParentID  IssueID
	DependsOn []IssueID
	RelatedTo []IssueID
	Blocked   bool
	Progress  Progress
	Cycle     int
	Phase     Phase
}

// An unsized issue scores 0 and sorts last: an estimate never made should not
// win by default.
func (i Issue) Score() float64 {
	if !i.Size.Valid() {
		return 0
	}
	weight := map[Priority]float64{PriorityLow: 1, PriorityMedium: 2, PriorityHigh: 3}[i.Priority]
	return weight / float64(i.Size)
}

type IssueFilter struct {
	Kind     IssueKind
	ParentID IssueID
	PlanID   PlanID
	Status   []IssueStatus
	Priority Priority
	Assignee UserID
	Tags     []string
	Search   string
	Limit    int
	Offset   int
}

type IssueBrief struct {
	Issue    Issue
	Children []Issue
	Plans    []Plan
	Journal  []JournalEntry
	Docs     []Doc
	Cycles   []Cycle
}

File: apps/web/src/core/domain/issue.ts
import type { Branded } from './branded';
import type { PlanId } from './plan';
import type { ProjectId, UserId } from './project';

export type IssueId = Branded<string, 'IssueId'>

export const issueId = (value: string): IssueId => value as IssueId

export const ISSUE_KIND = {
	TICKET: 'ticket',
	TODO: 'todo',
} as const

export const ISSUE_KINDS = [ISSUE_KIND.TICKET, ISSUE_KIND.TODO] as const

export type IssueKind = (typeof ISSUE_KINDS)[number]

export const ISSUE_STATUS = {
	OPEN: 'open',
	IN_PROGRESS: 'in_progress',
	BLOCKED: 'blocked',
	DONE: 'done',
	CANCELLED: 'cancelled',
} as const

export const ISSUE_STATUSES = [
	ISSUE_STATUS.OPEN,
	ISSUE_STATUS.IN_PROGRESS,
	ISSUE_STATUS.BLOCKED,
	ISSUE_STATUS.DONE,
	ISSUE_STATUS.CANCELLED,
] as const

export type IssueStatus = (typeof ISSUE_STATUSES)[number]

export const DEFAULT_TICKET_STATUSES: readonly IssueStatus[] = ISSUE_STATUSES.filter(status => status !== ISSUE_STATUS.DONE && status !== ISSUE_STATUS.CANCELLED)

export const PRIORITY = {
	LOW: 'low',
	MEDIUM: 'medium',
	HIGH: 'high',
} as const

export const PRIORITIES = [PRIORITY.LOW, PRIORITY.MEDIUM, PRIORITY.HIGH] as const

export type Priority = (typeof PRIORITIES)[number]

export const SIZES = [1, 2, 3, 5, 8] as const

export type Size = (typeof SIZES)[number]

export const WAYFINDER_TYPES = ['map', 'research', 'prototype', 'grilling', 'task'] as const

export type WayfinderType = (typeof WAYFINDER_TYPES)[number]

export const LINK_KIND = {
	BLOCKS: 'blocks',
	RELATES: 'relates',
	PARENT: 'parent',
} as const

export const LINK_KINDS = [LINK_KIND.BLOCKS, LINK_KIND.RELATES, LINK_KIND.PARENT] as const

export type LinkKind = (typeof LINK_KINDS)[number]

export type IssueLink = {
	readonly from: IssueId
	readonly to: IssueId
	readonly kind: LinkKind
}

export type Issue = {
	readonly id: IssueId
	readonly kind: IssueKind
	readonly projectId: ProjectId
	readonly planId: PlanId | undefined
	readonly slug: string
	readonly title: string
	readonly body: string
	readonly status: IssueStatus
	readonly priority: Priority
	readonly size: Size | undefined
	readonly assignee: UserId | undefined
	readonly tags: readonly string[]
	readonly position: number
	readonly dueDate: Date | undefined
	readonly wayfinder: WayfinderType | undefined
	readonly externalRef: string
	readonly createdBy: UserId | undefined
	readonly createdAt: Date
	readonly updatedAt: Date
	readonly parentId: IssueId | undefined
	readonly dependsOn: readonly IssueId[]
	readonly relatedTo: readonly IssueId[]
	readonly blocked: boolean
}

export const isTerminal = (status: IssueStatus): boolean => status === ISSUE_STATUS.DONE || status === ISSUE_STATUS.CANCELLED

const PRIORITY_WEIGHT: Record<Priority, number> = { low: 1, medium: 2, high: 3 }

export const score = (issue: Pick<Issue, 'priority' | 'size'>): number =>
	issue.size === undefined ? 0 : PRIORITY_WEIGHT[issue.priority] / issue.size

export const isSize = (value: number): value is Size => (SIZES as readonly number[]).includes(value)
